src/HazelcastClient.ts

```diff
@@ -440,18 +508,14 @@ export class HazelcastClient {
         this.statistics.stop();
         return this.cpSubsystem.shutdown()
             .then(() => {
-                this.connectionManager.shutdown();
                 this.invocationService.shutdown();
+                this.connectionManager.shutdown();
```

Comment:
I guess the test failure is due to this order change, I reverted in my local and I am trying to run the failing test 30 times in a script. If none of them fails, I can say that this is the issue. Before the change the fail was happening quite often like 1 in 5 trials.

Reply:
No luck, still failing

Reply:
Tried on windows, fails there too. Specifically this fails:

```
 1) ClientBackupAcksTest
       should receive backup acks in smart mode:
     Error: Invocation 8 failed because of missed backup acks
      at new IndeterminateOperationStateError (lib\core\HazelcastError.js:291:9)
      at Invocation.detectAndHandleBackupTimeout (lib\invocation\InvocationService.js:122:36)
      at Timeout._onTimeout (lib\invocation\InvocationService.js:202:32)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)
```

Any idea why?

Reply:
Could you try to enable `TRACE` log level and analyze the difference between successful and failed test run? If current logs are not enough, try to add more temporary logging into `InvocationService` and `Invocation` classes. It could be that the listener for backup acks (`InvocationService.backupEventHandler()`; also see `processResponse` method as it triggers the event handler) isn't registered or received in time. My bet would be that the flakiness is related with initialization order changes made in this PR, but I'm not 100% sure.

```js
await Client.newHazelcastClient({
    properties: {
        'hazelcast.logging.level': 'TRACE'
    }
});
```

There was also #693 which added a workaround for a member-side issue in this particular test, but as the test works reliable on `master`, it shouldn't be the reason.

Reply:
extra logs on fail 
```
[DefaultLogger] DEBUG at PartitionService: Handling new partition table with partitionStateVersion: 1
[DefaultLogger] TRACE at PartitionService: Event coming from a new connection. Old connection: undefined, new connection ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
```

full logs of a failing case, wrapped the extra lines with "~" chars:

```
[DefaultLogger] INFO at LifecycleService: HazelcastClient is STARTING
[DefaultLogger] INFO at LifecycleService: HazelcastClient is STARTED
[DefaultLogger] INFO at ConnectionManager: Trying to connect to 127.0.0.1:5701
[DefaultLogger] INFO at LifecycleService: HazelcastClient is CONNECTED
[DefaultLogger] INFO at ConnectionManager: Authenticated with server localhost:5701:1616a6ef-b167-49d0-bc3f-e35f8a2ffd23, server version: 4.1.1, local address: 127.0.0.1:33358
[DefaultLogger] TRACE at ClusterViewListenerService: Register attempt of cluster view handler to ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] TRACE at ClusterService: Resetting the member list version.
[DefaultLogger] INFO at ClusterService: 

Members [2] {
	Member [localhost]:5701 - 1616a6ef-b167-49d0-bc3f-e35f8a2ffd23
	Member [localhost]:5702 - fd4c8960-24ea-4f07-a482-e605f71d15b7
}

[DefaultLogger] TRACE at ClusterViewListenerService: Registered cluster view handler to ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] INFO at ConnectionManager: Authenticated with server localhost:5702:fd4c8960-24ea-4f07-a482-e605f71d15b7, server version: 4.1.1, local address: 127.0.0.1:59512
[DefaultLogger] DEBUG at ListenerService: Listener 54a75dbe-48c1-bc5c-586b-2d299801f708 registered on ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] DEBUG at ListenerService: Listener 54a75dbe-48c1-bc5c-586b-2d299801f708 registered on ClientConnection{alive=true, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] TRACE at InvocationService: Partition owner is not assigned yet
[DefaultLogger] TRACE at InvocationService: Partition owner is not assigned yet
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[DefaultLogger] DEBUG at PartitionService: Handling new partition table with partitionStateVersion: 1
[DefaultLogger] TRACE at PartitionService: Event coming from a new connection. Old connection: undefined, new connection ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[DefaultLogger] INFO at LifecycleService: HazelcastClient is SHUTTING_DOWN
[DefaultLogger] TRACE at Connection: ClientConnection{alive=false, connectionId=0, remoteAddress=localhost:5701} closed. Reason: Hazelcast client is shutting down
[DefaultLogger] INFO at ConnectionManager: Removed connection to endpoint: localhost:5701:1616a6ef-b167-49d0-bc3f-e35f8a2ffd23, connection: ClientConnection{alive=false, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] TRACE at ClusterViewListenerService: Register attempt of cluster view handler to ClientConnection{alive=true, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] TRACE at ClusterService: Resetting the member list version.
[DefaultLogger] TRACE at Connection: ClientConnection{alive=false, connectionId=1, remoteAddress=localhost:5702} closed. Reason: Hazelcast client is shutting down
[DefaultLogger] INFO at ConnectionManager: Removed connection to endpoint: localhost:5702:fd4c8960-24ea-4f07-a482-e605f71d15b7, connection: ClientConnection{alive=false, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] INFO at LifecycleService: HazelcastClient is DISCONNECTED
[DefaultLogger] INFO at LifecycleService: HazelcastClient is SHUTDOWN

Error: Invocation 8 failed because of missed backup acks
    at new IndeterminateOperationStateError (lib/core/HazelcastError.js:291:9)
    at Invocation.detectAndHandleBackupTimeout (lib/invocation/InvocationService.js:122:36)
    at Timeout._onTimeout (lib/invocation/InvocationService.js:202:32)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)

```

Still not sure what the problem is

Reply:
I don't think the issue is due to construction order. 

The only extra service added in this pr is connection registry. I reverted the services' construction order in a new branch: https://github.com/srknzl/hazelcast-nodejs-client/tree/test-reverting-contruction-order. To solve circular dependencies I used client object as in master.

I put the connection registry next to connection manager to make connection related initialization as much as similar to the one in master. The test is still failing.

Reply:
Another log comparison. I added 5 trace logs to these locations(first string in log.trace says the service name): 
![image](https://user-images.githubusercontent.com/32355782/109069421-42d7a200-7702-11eb-94a3-aee0efd4cb00.png)

![image](https://user-images.githubusercontent.com/32355782/109068551-15d6bf80-7701-11eb-96d8-8d84b6b9bf1e.png)

![image](https://user-images.githubusercontent.com/32355782/109069478-54b94500-7702-11eb-854c-6cba53a77369.png)

![image](https://user-images.githubusercontent.com/32355782/109069357-2a678780-7702-11eb-8fe5-b655bec53fe0.png)


success:
```
/home/serkan/node/bin/node /home/serkan/forked/hazelcast-nodejs-client/node_modules/mocha/bin/mocha --ui bdd --reporter /home/serkan/.local/share/JetBrains/WebStorm2020.3/NodeJS/js/mocha-intellij/lib/mochaIntellijReporter.js /home/serkan/forked/hazelcast-nodejs-client/test/integration/ClientBackupAcksTest.js --grep "^ClientBackupAcksTest should receive backup acks in smart mode$"
[DefaultLogger] INFO at LifecycleService: HazelcastClient is STARTING
[DefaultLogger] INFO at LifecycleService: HazelcastClient is STARTED
[DefaultLogger] INFO at ConnectionManager: Trying to connect to 127.0.0.1:5701
[DefaultLogger] INFO at LifecycleService: HazelcastClient is CONNECTED
[DefaultLogger] INFO at ConnectionManager: Authenticated with server localhost:5701:4f0b91d4-e90f-463e-bf8d-5864e0802a75, server version: 4.1.1, local address: 127.0.0.1:34736
[DefaultLogger] TRACE at clusterviewlistener: connectionadded
[DefaultLogger] TRACE at clusterviewlistener: tryRegister
[DefaultLogger] TRACE at ClusterViewListenerService: Register attempt of cluster view handler to ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] TRACE at ClusterService: Resetting the member list version.
[DefaultLogger] TRACE at invocation: processResponse
[DefaultLogger] INFO at ClusterService: 

Members [2] {
	Member [localhost]:5701 - 4f0b91d4-e90f-463e-bf8d-5864e0802a75
	Member [localhost]:5702 - 958b467e-de9c-4fc4-b513-af7116859b97
}

[DefaultLogger] TRACE at ClusterViewListenerService: Registered cluster view handler to ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] INFO at ConnectionManager: Authenticated with server localhost:5702:958b467e-de9c-4fc4-b513-af7116859b97, server version: 4.1.1, local address: 127.0.0.1:60890
[DefaultLogger] TRACE at clusterviewlistener: connectionadded
[DefaultLogger] TRACE at clusterviewlistener: tryRegister
[DefaultLogger] TRACE at listener: registerListener
[DefaultLogger] DEBUG at ListenerService: Listener a7b14624-1341-8c97-2002-0a90cfc430f5 registered on ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] DEBUG at ListenerService: Listener a7b14624-1341-8c97-2002-0a90cfc430f5 registered on ClientConnection{alive=true, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] TRACE at InvocationService: Partition owner is not assigned yet
[DefaultLogger] TRACE at InvocationService: Partition owner is not assigned yet
[DefaultLogger] TRACE at invocation: processResponse
[DefaultLogger] TRACE at invocation: backupEventHandler
[DefaultLogger] INFO at LifecycleService: HazelcastClient is SHUTTING_DOWN
[DefaultLogger] TRACE at Connection: ClientConnection{alive=false, connectionId=0, remoteAddress=localhost:5701} closed. Reason: Hazelcast client is shutting down
[DefaultLogger] INFO at ConnectionManager: Removed connection to endpoint: localhost:5701:4f0b91d4-e90f-463e-bf8d-5864e0802a75, connection: ClientConnection{alive=false, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] TRACE at clusterviewlistener: tryRegister
[DefaultLogger] TRACE at ClusterViewListenerService: Register attempt of cluster view handler to ClientConnection{alive=true, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] TRACE at ClusterService: Resetting the member list version.
[DefaultLogger] TRACE at Connection: ClientConnection{alive=false, connectionId=1, remoteAddress=localhost:5702} closed. Reason: Hazelcast client is shutting down
[DefaultLogger] INFO at ConnectionManager: Removed connection to endpoint: localhost:5702:958b467e-de9c-4fc4-b513-af7116859b97, connection: ClientConnection{alive=false, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] INFO at LifecycleService: HazelcastClient is DISCONNECTED
[DefaultLogger] INFO at LifecycleService: HazelcastClient is SHUTDOWN

```


fail:
```
[DefaultLogger] INFO at LifecycleService: HazelcastClient is STARTING
[DefaultLogger] INFO at LifecycleService: HazelcastClient is STARTED
[DefaultLogger] INFO at ConnectionManager: Trying to connect to 127.0.0.1:5701
[DefaultLogger] INFO at LifecycleService: HazelcastClient is CONNECTED
[DefaultLogger] INFO at ConnectionManager: Authenticated with server localhost:5701:d7ff7349-83fe-4cff-819c-0f248206f85f, server version: 4.1.1, local address: 127.0.0.1:34742
[DefaultLogger] TRACE at clusterviewlistener: connectionadded
[DefaultLogger] TRACE at clusterviewlistener: tryRegister
[DefaultLogger] TRACE at ClusterViewListenerService: Register attempt of cluster view handler to ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] TRACE at ClusterService: Resetting the member list version.
[DefaultLogger] TRACE at invocation: processResponse
[DefaultLogger] INFO at ClusterService: 

Members [2] {
	Member [localhost]:5701 - d7ff7349-83fe-4cff-819c-0f248206f85f
	Member [localhost]:5702 - 9b298be0-472b-4cda-aae3-f9943aad45b7
}

[DefaultLogger] TRACE at ClusterViewListenerService: Registered cluster view handler to ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] INFO at ConnectionManager: Authenticated with server localhost:5702:9b298be0-472b-4cda-aae3-f9943aad45b7, server version: 4.1.1, local address: 127.0.0.1:60896
[DefaultLogger] TRACE at clusterviewlistener: connectionadded
[DefaultLogger] TRACE at clusterviewlistener: tryRegister
[DefaultLogger] TRACE at listener: registerListener
[DefaultLogger] DEBUG at ListenerService: Listener 4f6ccf8d-5b53-b6e1-b667-c3fb0717f626 registered on ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] DEBUG at ListenerService: Listener 4f6ccf8d-5b53-b6e1-b667-c3fb0717f626 registered on ClientConnection{alive=true, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] TRACE at InvocationService: Partition owner is not assigned yet
[DefaultLogger] TRACE at InvocationService: Partition owner is not assigned yet
[DefaultLogger] TRACE at invocation: processResponse
[DefaultLogger] DEBUG at PartitionService: Handling new partition table with partitionStateVersion: 1
[DefaultLogger] TRACE at PartitionService: Event coming from a new connection. Old connection: undefined, new connection ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] INFO at LifecycleService: HazelcastClient is SHUTTING_DOWN
[DefaultLogger] TRACE at Connection: ClientConnection{alive=false, connectionId=0, remoteAddress=localhost:5701} closed. Reason: Hazelcast client is shutting down
[DefaultLogger] INFO at ConnectionManager: Removed connection to endpoint: localhost:5701:d7ff7349-83fe-4cff-819c-0f248206f85f, connection: ClientConnection{alive=false, connectionId=0, remoteAddress=localhost:5701}
[DefaultLogger] TRACE at clusterviewlistener: tryRegister
[DefaultLogger] TRACE at ClusterViewListenerService: Register attempt of cluster view handler to ClientConnection{alive=true, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] TRACE at ClusterService: Resetting the member list version.
[DefaultLogger] TRACE at Connection: ClientConnection{alive=false, connectionId=1, remoteAddress=localhost:5702} closed. Reason: Hazelcast client is shutting down
[DefaultLogger] INFO at ConnectionManager: Removed connection to endpoint: localhost:5702:9b298be0-472b-4cda-aae3-f9943aad45b7, connection: ClientConnection{alive=false, connectionId=1, remoteAddress=localhost:5702}
[DefaultLogger] INFO at LifecycleService: HazelcastClient is DISCONNECTED
[DefaultLogger] INFO at LifecycleService: HazelcastClient is SHUTDOWN

Error: Invocation 8 failed because of missed backup acks
    at new IndeterminateOperationStateError (lib/core/HazelcastError.js:291:9)
    at Invocation.detectAndHandleBackupTimeout (lib/invocation/InvocationService.js:122:36)
    at Timeout._onTimeout (lib/invocation/InvocationService.js:202:32)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)


```

Can't see what these mean

Reply:
This log message is unrelated with the issue. It simply means that the partition table wasn't received yet (which is ok; also `master` behaves in a similar way).
```
[DefaultLogger] DEBUG at PartitionService: Handling new partition table with partitionStateVersion: 1
[DefaultLogger] TRACE at PartitionService: Event coming from a new connection. Old connection: undefined, new connection ClientConnection{alive=true, connectionId=0, remoteAddress=localhost:5701}
```

Reply:
I've tried to reproduce the issue locally via running the test 15 times in a row. Got a single failure on this branch, so I'm able to reproduce it. Going to take a look at it.

Reply:
So far, I can see that in case of failure the same correlation id (`2`) is used for two different events (for both backup and cluster view listeners) which shouldn't happen. Going to look what may cause this problem.

Reply:
I enabled `FINEST` logging level in IMDG and it appears to be #616:
```
WARNING: [localhost]:5702 [6d5836ee-b431-4bbb-8050-ec61b238db4d] [4.2-SNAPSHOT] Wrong target! com.hazelcast.spi.impl.operationservice.impl.operations.Backup{serviceName='null', identityHash=522901856, partitionId=247, replicaIndex=1, callId=4, invocationTime=-1 (1970-01-01 02:59:59.999), waitTimeout=-1, callTimeout=9223372036854775807, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0, backupOp=com.hazelcast.map.impl.operation.PutBackupOperation{serviceName='hz:impl:mapService', identityHash=454381149, partitionId=247, replicaIndex=1, callId=0, invocationTime=1614252757826 (2021-02-25 14:32:37.826), waitTimeout=-1, callTimeout=9223372036854775807, tenantControl=com.hazelcast.spi.impl.tenantcontrol.NoopTenantControl@0, name=test-map}, backupOpData=null, originalCaller=[localhost]:5701, version=[1, 0, 0, 0, 0, 0], sync=true} cannot be processed! Target should be: null
```

Reopening that issue and we can ignore flakiness of `ClientBackupAcksTest` for this PR.

Obviously #693 didn't fix the issue reliably. As a more reliable fix we could wait for the partition table event to be received by the client and only then send the `map.set()` operation.

cc @mdumandag

Reply:
How about master not facing this issue

Reply:
The fail of `ClientBackupAcksTest` was due to a bug that I realized after reviewing all of my changes again. Connection registry was not initialized correctly (specifically `loadBalancer` and `smartRouting` attributes was not being set in the constructor), and because I did not set `smartRouting` variable, load balancer is not used in `getRandomConnection` resulting no error.

You can see the change in  https://github.com/srknzl/hazelcast-nodejs-client/commit/4cade713d4ab3bfeb67ec4d61c3821d153eb2618

I ran the clientbackup test with this loop for 12 minutes:

```python
from subprocess import run

while True:
    p = run(['npx', 'mocha', '-g', '"should receive backup acks in smart mode"', '--recursive', 'test'])
    if p.returncode != 0:
        print(p.stderr)
        print(p.stdout)
        print(p.returncode)
        break
```
It did not fail, I think the change solved the issue.

Additionally, 
I ran the other test(`HeartbeatFromServerTest`) on its own on this PR's branch(https://github.com/srknzl/hazelcast-nodejs-client/commit/acdefe3fe6a47f4a308310d6e0b5731dc9db60a5) and on master(https://github.com/hazelcast/hazelcast-nodejs-client/commit/9e7cd0360ea4769edb8dcaba8cc3f40daec96486). Both failed several minutes later with the same error message:
```
1) HeartbeatFromServerTest
       connectionRemoved fired when second member stops heartbeating:
     Error: Timeout of 120000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves. (/home/serkan/git/hazelcast-nodejs-client/test/integration/heartbeat/HeartbeatFromServerTest.js)
      at listOnTimeout (internal/timers.js:554:17)
      at processTimers (internal/timers.js:497:7)

```
So I don't think this is a problem specific to this branch.

Just to be sure every test still passes, I ran the whole unit+integration test suit 3 times. All passed.

Reopening the issue #616's was most probably a false alarm. @mdumandag  

With this issue fixed, I suppose there is no issue left blocking merge of this pr.

Reply:
> Connection registry was not initialized correctly (specifically `loadBalancer` and `smartRouting` attributes was not being set in the constructor), and because I did not set `smartRouting` variable, load balancer is not used in `getRandomConnection` resulting no error.

That's a good catch and surely it was a bug. However, this doesn't change the fact that the issue is member-side. I'd prefer to leave #616 open for now as it's very likely to show up in the future unless we make sure to use the [suggested workaround](https://github.com/hazelcast/hazelcast-nodejs-client/pull/797#discussion_r582762358).

The bugfix you've made affects client-to-member comms in the following aspect: before your fix load balancer wasn't in place and smart client was sending all of its requests to the very first member from the active connection list (just like unisocket client does it). With your change it started to send requests in a round robin fashion which helped to mitigate the issue. If in future the client behavior in terms of load balancer or active connections list handling changes, the issue may show up once again.

Also, it would be great to have a simple unit test which validates `getRandomConnection` in smart/unisocket modes. Created #807 to track this enhancement.